lib/src/parser/recursive.dart

```diff
@@ -253,39 +257,56 @@ class RecursiveAstParser {
 
     // If not a void element, look for closing tag OR child nodes.
     if (!isVoidElement && nextToken.type != NgTokenType.openElementEndVoid) {
-      // Collect child nodes.
-      nextToken = _reader.next();
-      // Guaranteed by scanner w/ recovery that next token is beginning of a
-      // a standaloneAst-starting token OR null.
-      while (nextToken != null &&
-          nextToken.type != NgTokenType.closeElementStart) {
-        TemplateAst childAst = parseStandalone(nextToken);
-        childNodes.add(childAst);
+      tagStack.addFirst(nameToken.lexeme);
+      var closingTagFound = false;
+
+      while (!closingTagFound) {
         nextToken = _reader.next();
-      }
-      if (nextToken == null) {
-        exceptionHandler.handle(new AngularParserException(
-          "Expected close element for '${nameToken.lexeme}'",
-          nameToken.lexeme,
-          nameToken.offset,
-        ));
-        closeElementAst = new CloseElementAst(nameToken.lexeme);
-      } else {
-        final closeNameToken = _reader.peek();
-        // Found a close tag, but unmatching identifier.
-        if (closeNameToken.lexeme != nameToken.lexeme) {
+
+        if (nextToken == null) {
           exceptionHandler.handle(new AngularParserException(
-            'Invalid closing tag: $closeNameToken (expected $nameToken)',
-            closeNameToken.lexeme,
-            closeNameToken.offset,
+            "Expected close element for '${nameToken.lexeme}'",
+            nameToken.lexeme,
+            nameToken.offset,
           ));
-          childNodes.add(_handleDanglingCloseElement(nextToken));
           closeElementAst = new CloseElementAst(nameToken.lexeme);
-          // Correct path
+          closingTagFound = true;
+        } else if (nextToken.type == NgTokenType.closeElementStart) {
+          var closeNameToken = _reader.peek();
+
+          if (closeNameToken.lexeme != nameToken.lexeme) {
+            // Found a closing tag, but not matching current [ElementAst].
+            exceptionHandler.handle(new AngularParserException(
+              'Invalid closing tag: $closeNameToken (expected $nameToken)',
+              closeNameToken.lexeme,
+              closeNameToken.offset,
+            ));
+            if (tagStack.contains(closeNameToken.lexeme)) {
```

Comment:
It would be cool to add a 'threshold' to see when to hold off in terms of stack length. For example, if it exceeds 2 stacks, it treats it as a dangling - in which case it would be resolved as:

<div><ul><li><span></span><div></div></li></ul></div>, but this might have issues within it as well. As you mentioned, adding a threshold would prevent this from triggering too often and completely mangling the original.